File: solutions/python/06/mo/main.py
def mo_order(queries, block):
    ordered = []
    for idx, (l, r) in enumerate(queries):
        bid = l // block
        ordered.append((bid, r if bid % 2 else -r, l, r, idx))
    ordered.sort()
    return ordered


def mo_count_distinct(a, queries):
    n = len(a)
    block = int(n**0.5) + 1
    order = mo_order(queries, block)
    cnt = {}
    cur_l, cur_r = 0, 0
    distinct = 0
    ans = [0] * len(queries)

    def add(i):
        nonlocal distinct
        x = a[i]
        cnt[x] = cnt.get(x, 0) + 1
        if cnt[x] == 1:
            distinct += 1

    def remove(i):
        nonlocal distinct
        x = a[i]
        cnt[x] -= 1
        if cnt[x] == 0:
            distinct -= 1

    for _, _, l, r, idx in order:
        while cur_r <= r:
            add(cur_r)
            cur_r += 1
        while cur_r - 1 > r:
            cur_r -= 1
            remove(cur_r)
        while cur_l > l:
            cur_l -= 1
            add(cur_l)
        while cur_l < l:
            remove(cur_l)
            cur_l += 1
        ans[idx] = distinct
    return ans


def main() -> None:
    a = [1, 2, 1, 3, 2, 1, 2]
    queries = [(0, 6), (2, 5), (1, 4)]
    print(mo_count_distinct(a, queries))


if __name__ == "__main__":
    main()

File: solutions/go/06/mo/main.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func main() {
	a := []int{1, 2, 1, 3, 2, 1, 2}
	queries := [][2]int{{0, 6}, {2, 5}, {1, 4}}
	n := len(a)
	block := int(math.Sqrt(float64(n))) + 1
	type item struct {
		bid, r, l, idx int
	}
	items := make([]item, len(queries))
	for i, q := range queries {
		bid := q[0] / block
		r := q[1]
		if bid%2 == 1 {
			r = -r
		}
		items[i] = item{bid, r, q[0], i}
	}
	sort.Slice(items, func(i, j int) bool {
		if items[i].bid != items[j].bid {
			return items[i].bid < items[j].bid
		}
		return items[i].r < items[j].r
	})
	cnt := map[int]int{}
	curL, curR := 0, 0
	distinct := 0
	ans := make([]int, len(queries))
	add := func(i int) {
		x := a[i]
		cnt[x]++
		if cnt[x] == 1 {
			distinct++
		}
	}
	remove := func(i int) {
		x := a[i]
		cnt[x]--
		if cnt[x] == 0 {
			distinct--
		}
	}
	for _, it := range items {
		l, r := it.l, queries[it.idx][1]
		for curR <= r {
			add(curR)
			curR++
		}
		for curR-1 > r {
			curR--
			remove(curR)
		}
		for curL > l {
			curL--
			add(curL)
		}
		for curL < l {
			remove(curL)
			curL++
		}
		ans[it.idx] = distinct
	}
	fmt.Println(ans)
}
